File: util/java/src/main/java/com/dmi/util/text/CharSequenceCharacterIterator.java
package com.dmi.util.text;

import java.text.CharacterIterator;

import static com.google.common.base.Preconditions.checkArgument;

public class CharSequenceCharacterIterator implements CharacterIterator {
    private final CharSequence text;

    private int index;

    public CharSequenceCharacterIterator(CharSequence text) {
        this.text = text;
        index = 0;
    }

    @Override
    public int getBeginIndex() {
        return 0;
    }

    @Override
    public int getEndIndex() {
        return text.length();
    }

    @Override
    public int getIndex() {
        return index;
    }

    @Override
    public char current() {
        return index < text.length() ? text.charAt(index) : DONE;
    }

    @Override
    public char first() {
        index = 0;
        return current();
    }

    @Override
    public char last() {
        index = text.length() > 0 ? text.length() - 1 : 0;
        return current();
    }

    @Override
    public char next() {
        return index < text.length() - 1 ? text.charAt(++index) : DONE;
    }

    @Override
    public char previous() {
        return index > 0 ? text.charAt(--index) : DONE;
    }

    @Override
    public char setIndex(int index) {
        checkArgument(index >= 0 || index < text.length());
        this.index = index;
        return current();
    }

    public Object clone() {
        try {
            return super.clone();
        } catch (CloneNotSupportedException e) {
            throw new RuntimeException(e);
        }
    }
}

File: util/java/src/main/kotlin/com/dmi/util/text/CharSequenceCharacterIterator.kt
package com.dmi.util.text

import com.google.common.base.Preconditions.checkArgument
import java.text.CharacterIterator

class CharSequenceCharacterIterator(private val text: CharSequence) : CharacterIterator {
    private var index: Int = 0

    init {
        index = 0
    }

    override fun getBeginIndex(): Int {
        return 0
    }

    override fun getEndIndex(): Int {
        return text.length
    }

    override fun getIndex(): Int {
        return index
    }

    override fun current(): Char {
        return if (index < text.length) text[index] else CharacterIterator.DONE
    }

    override fun first(): Char {
        index = 0
        return current()
    }

    override fun last(): Char {
        index = if (text.length > 0) text.length - 1 else 0
        return current()
    }

    override fun next(): Char {
        return if (index < text.length - 1) text[++index] else CharacterIterator.DONE
    }

    override fun previous(): Char {
        return if (index > 0) text[--index] else CharacterIterator.DONE
    }

    override fun setIndex(index: Int): Char {
        checkArgument(index >= 0 || index < text.length)
        this.index = index
        return current()
    }

    override fun clone(): Any {
        val clone = CharSequenceCharacterIterator(text)
        clone.index = index
        return clone
    }
}
